packages/logging/src/logging-and-telemetry.ts

```diff
@@ -0,0 +1,791 @@
+import redactInfo from 'mongodb-redact';
+import { redactURICredentials } from '@mongosh/history';
+import type {
+  MongoshBus,
+  ApiEventWithArguments,
+  ApiEvent,
+  UseEvent,
+  EvaluateInputEvent,
+  ShowEvent,
+  ConnectEvent,
+  ScriptLoadFileEvent,
+  StartLoadingCliScriptsEvent,
+  StartMongoshReplEvent,
+  GlobalConfigFileLoadEvent,
+  CryptLibrarySkipEvent,
+  CryptLibraryFoundEvent,
+  SnippetsCommandEvent,
+  SnippetsErrorEvent,
+  SnippetsFetchIndexErrorEvent,
+  SnippetsFetchIndexEvent,
+  SnippetsLoadedEvent,
+  SnippetsLoadSnippetEvent,
+  SnippetsNpmDownloadActiveEvent,
+  SnippetsNpmDownloadFailedEvent,
+  SnippetsNpmLookupEvent,
+  SnippetsRunNpmEvent,
+  SnippetsTransformErrorEvent,
+  EditorRunEditCommandEvent,
+  EditorReadVscodeExtensionsDoneEvent,
+  EditorReadVscodeExtensionsFailedEvent,
+  FetchingUpdateMetadataEvent,
+  FetchingUpdateMetadataCompleteEvent,
+  SessionStartedEvent,
+  MongoshBusEventsMap,
+} from '@mongosh/types';
+import { inspect } from 'util';
+import { MongoLogWriter } from 'mongodb-log-writer';
+import { mongoLogId } from 'mongodb-log-writer';
+import type {
+  AnalyticsIdentifyMessage,
+  MongoshAnalytics,
+  MongoshAnalyticsIdentity,
+} from './analytics-helpers';
+import { hookLogger } from '@mongodb-js/devtools-connect';
+import { MultiSet, toSnakeCase } from './helpers';
+import { Writable } from 'stream';
+
+export class MongoshLoggingAndTelemetry {
+  public log: MongoLogWriter;
+
+  private readonly trackingProperties: {
+    mongosh_version: string;
+    session_id: string;
+  };
+  private readonly userTraits: AnalyticsIdentifyMessage['traits'] & {
+    [key: string]: unknown;
+  };
+
+  private pendingLogEvents: CallableFunction[] = [];
+
+  private telemetryAnonymousId: string | undefined;
+  private userId: string | undefined;
+  private hasMongoLogWriterInitialized = false;
+  private isSetup = false;
+
+  constructor(
+    public readonly bus: MongoshBus,
+    public readonly analytics: MongoshAnalytics,
+    providedTraits: { platform: string } & {
+      [key: string]: unknown;
+    },
+    mongoshVersion: string
+  ) {
+    const dummySink = new Writable({
+      write(chunk, encoding, callback) {
+        callback();
+      },
+    });
+    this.log = new MongoLogWriter('', null, dummySink);
+
+    this.userTraits = {
+      ...providedTraits,
+      session_id: this.log.logId,
+    };
+
+    this.trackingProperties = {
+      mongosh_version: mongoshVersion,
+      session_id: this.log.logId,
+    };
+  }
+
+  public setup(): void {
+    if (this.isSetup) {
+      throw new Error('Setup can only be called once.');
+    }
+    this._setupBusEventListeners();
+    this.isSetup = true;
+  }
+
+  public setupLogger(logger: MongoLogWriter): void {
```

Comment:
This might be too destructive, I could limit it to only event defined here but then we don't end up handling `hookLogger` things